WIP: New SciPy backend

phschiele opened the pull request:
## Description
This PR attempts to bring a new backend to CVXPY that no longer relies on CVXCORE.
I.e., the code in `lin_ops/parser.py` should contain all functionality currently in the C++ code.
Preliminary testing shows that it is usually faster when no/few parameters are involved. 
For problems with many parameters, the stacking of 2D sparse matrices is slow. A follow-up with ND-sparse libraries is in progress.
The main idea is to make this feature `opt-in` now to get more feedback on its performance in different applications.
Issue link (if applicable): #1521

Some questions @cvxpy/maintainers:
1. What's the best way to pipe the backend choice through? Obviously, the current implementation is just for demonstration
2. What about testing? For development, I added a special flag that would test all backends and ensure the result is the same across backends, but I was not sure if we would like to have such a flag exposed.

TODOs:
- [x] Documentation
- [x] Remove some code duplication
- [x] Unit tests
- [x] Regression tests (i.e., running all tests under the new backend)

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [x] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [x] Run the benchmarks to make sure your change doesn’t introduce a regression.

SteveDiamond commented:
I would expose the backend similarly to how we expose solvers, as an argument to solve.

Maybe you could add another test file that focuses on comparing backends?

SteveDiamond commented on cvxpy/lin_ops/parser.py:
```diff
@@ -0,0 +1,743 @@
+    def get_A_b(self):
+        """
+        Returns [A b]
+        """
```
should this be ``pass``?

phschiele replied:
Done

phschiele replied:
Actually, @SteveDiamond, this was automatically removed by `autoflake`, because the function has a docstring.
Flake8 itself did not complain, however, so we might reconsider using that tool.

SteveDiamond replied:
yeah, maybe best to remove autoflake.

phschiele replied:
Added a `# noqa` for now to not pollute this PR with a potential removal of autoflake.

SteveDiamond commented:
The algorithm looks fine to me. The code is very concise, cleaner but still similar to what's in cvxcore. I wonder if we should run all the tests against each of the backends rather than writing new tests that explicitly vary the backend?

phschiele commented:
@SteveDiamond We can have the full python/os matrix run on the default backend, and have one config run the other backends.

phschiele commented on cvxpy/lin_ops/parser.py:
```diff
@@ -0,0 +1,743 @@
+    def get_A_b(self):
+        """
+        Returns [A b]
+        """
```
```suggestion
        """
        pass
```

phschiele commented:
For problems not containing any parameters, the improvements are:
cvar_benchmark: 15%
optimal_advertising: 74%
sdp_segfault_1132_benchmark: 85%
SimpleLPBenchmark: 5%

rileyjmurray commented on cvxpy/lin_ops/parser.py:
```diff
@@ -0,0 +1,776 @@
+    data: np.ndarray
+
+    @classmethod
+    def combine(cls, tensors: list[TensorRepresentation]):
+        parameter_offset, row, col, data = np.array([]), np.array([]), np.array([]), np.array([])
+        for t in tensors:
+            parameter_offset = np.append(parameter_offset, t.parameter_offset)
+            row = np.append(row, t.row)
+            col = np.append(col, t.col)
+            data = np.append(data, t.data)
+        return cls(parameter_offset, row, col, data)
```
It might be faster to use lists instead of numpy arrays. A numpy array needs to be contiguous in-memory, while a list doesn't have that requirement. (Of course you'd convert the lists to arrays before returning.)

phschiele replied:
@rileyjmurray Great spot! I had written as you described initially for this very reason, but did some benchmarks as it became one of the bottlenecks. To my surprise, the above method was the fasted one in my testing.
I used actual problem data back then, but here is an example of what I tested:
```py
import numpy as np
A = [np.arange(i,i+1000) for i in range(20)]
```
Lists:
```py
%%timeit
res = []
for a in A:
    res.extend(a)
np.array(res)  
>>> 1.29 ms ± 14.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
`np.append`
```py
%%timeit
res = np.array([])
for a in A:
    np.append(res, a)
>>> 76.3 µs ± 181 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
Perhaps a comment could be added that, for our use case, this is faster than using lists?

rileyjmurray commented on cvxpy/lin_ops/parser.py:
```diff
@@ -0,0 +1,776 @@
+            offset += np.prod(arg.shape)
+            if res is None:
+                res = arg_view
+            else:
+                res += arg_view
+        assert res is not None
+        return res
+
+    def vstack(self, lin: LinOp, view: TensorView) -> TensorView:
+        view = self.hstack(lin, view)
+        offset = 0
+        indices = []
+        for arg in lin.args:
+            arg_rows = np.prod(arg.shape)
+            indices.append(np.arange(arg_rows).reshape(arg.shape, order="F") + offset)
+            offset += arg_rows
+        order = np.vstack(indices).flatten(order="F").astype(int)
+        view.select_rows(order)
+        return view
+
+    @staticmethod
+    def transpose(lin: LinOp, view: TensorView) -> TensorView:
+        rows = np.arange(np.prod(lin.shape)).reshape(lin.shape).flatten(order="F")
+        view.select_rows(rows)
+        return view
+
+    @staticmethod
+    def upper_tri(lin: LinOp, view: TensorView) -> TensorView:
+        indices = np.arange(np.prod(lin.args[0].shape)).reshape(lin.args[0].shape, order="F")
+        triu_indices = indices[np.triu_indices_from(indices, k=1)]
+        view.select_rows(triu_indices)
+        return view
+
+    @staticmethod
+    def diag_mat(lin: LinOp, view: TensorView) -> TensorView:
+        # diagonal matrix to vector
+        rows = lin.shape[0]
+        diag_indices = np.arange(rows) * rows + np.arange(rows)
+        view.select_rows(diag_indices)
+        return view
```
Are these implementations supposed to be used for "all" backends?

phschiele replied:
Precisely, given `select_rows` is an abstract method that all backends should support, this approach resulted in the least duplication (when I added additional backends locally).

phschiele replied:
@rileyjmurray Take a look at cfef17d, specifically `test_transpose()`, `test_upper_tri()`, and `test_diag_mat()`. They all use the row selection logic from the base class, so the SciPy backend only has to implement `select_rows`.
I will add a shortened version of the test description as a doc string for the functions.

rileyjmurray replied:
Wonderful! Thanks for clarifying.

rileyjmurray commented on cvxpy/lin_ops/parser.py:
```diff
@@ -0,0 +1,776 @@
+class ScipyBackend(Backend):
+
+    @staticmethod
+    def reshape_constant_data(constant_data: dict[int, sp.csr_matrix], new_shape: tuple[int]):
+        return {k: [v_i.reshape(new_shape[1:], order="F")
+                    for v_i in v] for k, v in constant_data.items()}
```
Can this be made more "flat" instead of nested?

rileyjmurray replied:
To clarify: I'd like to be able to use the debugger in this function. Right now I can't set any useful breakpoints. Having explicit for-loops makes it much easier to debug.

phschiele replied:
I tried to use dict / list comprehensions due to their slight performance benefits. I agree that this makes it slightly harder to debug, though. I did set a breakpoint at the return statement and called `next()` on the dict to get the first key, value pair. Then I applied the inner function to the first element of the value (list) for debugging. One can also quickly break up the comprehensions for debugging, if necessary. 
Since performance is a major consideration for this part of the code, I would suggest keeping the comprehensions, even if it makes debugging slightly more cumbersome. If you have a strong view otherwise, we can also change it.

rileyjmurray commented on cvxpy/lin_ops/parser.py:
```diff
@@ -0,0 +1,776 @@
+    def reshape_tensors(self, tensor: TensorRepresentation, total_rows: int) -> sp.csc_matrix:
+        # Windows uses int32 by default at time of writing, so we need to enforce int64 here
+        rows = (tensor.col.astype(np.int64) * np.int64(total_rows) + tensor.row.astype(np.int64))
+        cols = tensor.parameter_offset.astype(int)
```
Is there a reason why ``rows`` needs to be ``int64`` but ``cols`` can just be ``int``?

phschiele replied:
Fair point, I only changed this after tests failed. Generally, I try using native python types.
This only failed under windows btw, so I had to be specific here([reason](https://stackoverflow.com/questions/36278590/numpy-array-dtype-is-coming-as-int32-by-default-in-a-windows-10-64-bit-machine)).
`rows` is of the order `A.size`, whereas `cols` is of the order `parameters.size`.
I will change it to be consistent within this function, though.

phschiele replied:
Done.

rileyjmurray commented on cvxpy/lin_ops/parser.py:
```diff
@@ -0,0 +1,776 @@
+    @abstractmethod
+    def concatenate_tensors(self, tensors: list[tuple[TensorView, int]]) -> TensorView:
+        """
+        Takes list of tensors and stacks them "row-wise"
```
What's meant by "row-wise"? If I have tensors of shape ``(x, y, z)`` and ``(p, q, ell)``, do we require ``(y, z) == (q, ell)`` and return a tensor of shape ``(x + p, y, z)``?

phschiele replied:
I have clarified row-wise to mean "axis 0". Your example is exactly what this function does, so it aligns with familiar concatenation functions from numpy, etc. .

rileyjmurray commented on cvxpy/problems/problem.py:
```diff
@@ -610,6 +610,11 @@ def get_problem_data(
             Raised if DPP settings are invalid.
         """
+
+        # TODO: remove
```
Please make this TODO clearer.

phschiele replied:
TODO has been removed.

rileyjmurray reviewed (changes requested):
I left a few specific questions. Overall, I think it's important that nontrivial documentation be written for these functions. I'll ask that classes and their main methods to be documented at a similar level of reductions in this file: https://github.com/cvxpy/cvxpy/blob/master/cvxpy/reductions/cone2cone/affine2direct.py. In my ideal world the individual functions like ``ScipyBackend.mul`` would describe the mathematical form of their output.

akshayka commented on cvxpy/lin_ops/canon_backend.py:
```diff
@@ -0,0 +1,785 @@
+        return cls(parameter_offset, row, col, data)
+
+
+class CanonBackend(ABC):
```
Can you add a short docstring that describes what functionality this class provides, and explains the semantics of its attributes?

phschiele replied:
Added doc strings to all functions in the file.

akshayka commented on cvxpy/lin_ops/canon_backend.py:
```diff
@@ -0,0 +1,785 @@
+        pass
+
+    @abstractmethod
+    def get_variable_tensor(self, shape: tuple[int], variable_id: int):
```
Can you document this method, as well as `get_data_tensor` and `get_param_tensor`.

phschiele replied:
Done.

akshayka commented on cvxpy/lin_ops/canon_backend.py:
```diff
@@ -0,0 +1,785 @@
+        return {parameter_id: slices}
+
+
+class TensorView(ABC):
```
Document the functionality of this class and the semantics of its attributes, please.

phschiele replied:
Done.

akshayka commented on cvxpy/lin_ops/canon_backend.py:
```diff
@@ -0,0 +1,785 @@
+
+        self.apply_all(func)
+
+    def apply_all(self, func: Callable) -> None:
```
Can you document the arguments to and semantics of this method, as well as `accumulate_over_variables` and `apply_to_parameters`?

phschiele replied:
Done.

akshayka commented on cvxpy/lin_ops/canon_backend.py:
```diff
@@ -0,0 +1,785 @@
+                               self.id_to_col, self.param_to_size, self.param_to_col,
+                               self.var_length)
+
+    def accumulate_over_variables(self, func, is_param_free_function: bool):
```
Are this method and the function `apply_to_parameters` specific to `ScipyTensorView`? I noticed that they are not declared as members of `TensorView`.

phschiele replied:
Great spot. Indeed, this function is not called by any superclass function, so I did not add it as an abstractmethod. 
For example, I have a local experimental "NumpyTensorView" (for debugging purposes only), which does not have this function.

akshayka commented on cvxpy/lin_ops/canon_backend.py:
```diff
@@ -0,0 +1,785 @@
+            return ScipyTensorView.add_dicts(a, b)
+
+    @staticmethod
+    def add_dicts(a, b):
```
Docstring, and add type annotations?

phschiele replied:
Done.

akshayka commented on cvxpy/lin_ops/canon_backend.py:
```diff
@@ -0,0 +1,785 @@
+
+
+class TensorView(ABC):
+    def __init__(self, variable_ids: set[int] | None, tensor, is_parameter_free: bool,
```
`tensor` is missing type annotation

phschiele replied:
Thanks, added it.

akshayka reviewed (commented):
This looks great! Like Riley, I've asked for documentation in a few key places. It's great that functions and methods have type annotations, though I noticed type annotations are missing in just a few places; it would help legibility to annotate all arguments and return types.

akshayka commented:
> I would expose the backend similarly to how we expose solvers, as an argument to solve.

If we do this, we should have a plan to deprecate the `CPP` backend in the future. It would be painful to have to maintain two implementations of `build_matrix`.

SteveDiamond commented:
> > I would expose the backend similarly to how we expose solvers, as an argument to solve.
> 
> If we do this, we should have a plan to deprecate the `CPP` backend in the future. It would be painful to have to maintain two implementations of `build_matrix`.

The two backends have different performance characteristics so unfortunately we have to roll with both for a while. It's definitely not possible to switch the default backend without much more testing.

phschiele commented:
The SDP example is significantly slower under the new backend, even though it does not use parameters.
This is worth looking into, but as the current suggestion is to add the new backend as experimental/non-default, this should not be a blocker.

```
 before           after         ratio
     [d15197bf]       [e3576451]
+      7.27±0s          41.3±0s     5.69  semidefinite_programming.SemidefiniteProgramming.time_compile_problem
```

Thanks @parthb83 for this recent addition to the benchmark suite!

phschiele commented on cvxpy/lin_ops/canon_backend.py:
```diff
@@ -0,0 +1,785 @@
+    data: np.ndarray
+
+    @classmethod
+    def combine(cls, tensors: list[TensorRepresentation]):
```
This is possible now, thanks to `from __future__ import annotations` (i.e. adding the return type `TensorRepresentation` within the class `TensorRepresentation`.

phschiele replied:
Done.

rileyjmurray commented on cvxpy/problems/problem.py:
```diff
@@ -589,6 +591,10 @@ def get_problem_data(
             When True, DPP problems will be treated as non-DPP,
             which may speed up compilation. Defaults to False.
+        canon_backend : str, optional
+            specifies which backend to use for canonicalization, which can affect
+            performance. Defaults to None, i.e., selecting the default
```
Sounds good to me!

phschiele replied:
Done.

phschiele commented on cvxpy/problems/problem.py:
```diff
@@ -940,6 +952,10 @@ def _solve(self,
+        canon_backend : str, optional
+            specifies which backend to use for canonicalization, which can affect
+            performance. Defaults to None, i.e., selecting the default
+            backend.
```
Done.

phschiele commented:
@cvxpy/maintainers Many thanks for your detailed reviews! I have addressed all of your comments above. 
In particular, I have added:
- Extensive tests, with unit tests covering all major functions of the backend, and regression tests covering every line containing logic. Note: the coverage will not show correctly on sonar, as the coverage run uses the CPP backend.
- A new workflow that runs all tests on the new backend. This workflow is not executed as part of the PR yet, so we only see if it works as expected when we merge and potentially need to readjust if it doesn't.
- All functions have short docstrings and complete type annotations now. The docstrings are always in the base class to not repeat them for each subclass. They should get passed down on most IDEs e.g. when looking up a function.
- Added a very short web documentation section just stating the semantics. We can expand this and give more details if you think it makes sense. We can also consider adding it to the "News" section of the docs, since the performance impact can be major.

SteveDiamond commented on doc/source/tutorial/advanced/index.rst:
```diff
@@ -1581,3 +1581,14 @@ you should set the class variable ``MIP_CAPABLE`` to ``True``. If your solver is
+Canonicalization backends
+------------------------------------
+Users can select from multiple canonicalization backends by adding the ``canon_backend``
+keyword argument to the ``.solve()`` call, e.g. ``problem.solve(canon_backend="SCIPY")`` (Introduced in CVXPY 1.3).
```
Usually we add a global in settings.py for any string like "SCIPY"

phschiele replied:
I did, and I use that string internally as `s.SCIPY_CANON_BACKEND`. I will also add it to the top level `__init__` to be consistent with the solver options.

phschiele replied:
Added here [118ee3c](https://github.com/cvxpy/cvxpy/pull/1840/commits/118ee3c5e1cd39effabe7f0715b3e7a2a56d7f7b)

SteveDiamond commented:
Thanks for wrapping this up! Definitely put something in the news section!

akshayka commented:
Looks great! What a Herculean PR. Congrats Philipp for getting this through.

SteveDiamond commented:
Thanks @phschiele !!!! This is going to make a huge difference for cvxpy users, and in general open up a whole new area of investigation. I looked through everything again and it all looks great! The tests are very helpful for documenting the details of how all the different linops are handled.
